### DISCUSSION TITLE
If it is possible to introduce materials effects such as programming noise during update

### DISCUSSION THREAD

**Discussion by ZhenyuWu323** [IBM/aihwkit#552] [Q&A]
If it is possible to introduce materials effects such as device noise during update, such as adding noises during gradient updates?  If so, where can I modify or access the device noise, through devices defined in `devices.py`? If I can implement this update behavior through modify the RPUConfigs defined in `configs.py` or I will probably need to implement a custom optimizer through deriving a new class from an AnalogTile or CustomTile and overriding that method?

**Comment by maljoras** [comment]
Certainly, update device noise is possible to simulate. We have a wide-range of device models available. The RPUConfig defines what device model and what update behavior is used. For instance, we have a version of "mixed precision" where the gradients are accumulated in digital and then written (with device noise) onto the (analog) weights. Or the gradient can also be accumulated directly onto analog using stochastic pulse trains. I would suggested to look at the [tutorial paper](https://arxiv.org/abs/2307.09357), which describes these ideas in mode detail (Table III, Section V).

**Reply by ZhenyuWu323** [reply]
Hi, Thank you for the previous info, and that tutorial paper really help me a lot! One more confusion. RPUConfig defines what device model and what update behavior is used, and we can apply different update algorithm like the "mixed precision" and  "Tiki-taka" through Compounds that are used to define the update bahavior of RPUConig (Table IX of the tutorial paper). However, it also seems that the in-memory update behavior is also defined in `analog_tile.update` method called by the `AnalogOptimizer`. Where those  in-memory training algorithm like "mixed precision" were actually implemented(It seems` Compounds.py` doesn't implement the algorithm itself), and how they will be called during in-memory training? What is the difference between them two?
